examples/gno.land/p/gnoland/commondao/proposal.gno

```diff
@@ -0,0 +1,174 @@
+package commondao
+
+import (
+	"errors"
+	"std"
+	"time"
+)
+
+// DefaultQuorum defines the default quorum required to tally proposal votes.
+const DefaultQuorum = 0.34 // 34%
+
+const (
+	StatusActive   ProposalStatus = "active"
+	StatusFailed                  = "failed"
+	StatusExecuted                = "executed"
+)
+
+const (
+	ChoiceAbstain VoteChoice = ""
+	ChoiceYes                = "YES"
+	ChoiceNo                 = "NO"
+)
+
+var (
+	ErrInvalidCreatorAddress      = errors.New("invalid proposal creator address")
+	ErrProposalDefinitionRequired = errors.New("proposal definition is required")
+	ErrStatusIsNotActive          = errors.New("proposal status is not active")
+)
+
+type (
+	// ProposalStatus defines a type for different proposal states.
+	ProposalStatus string
+
+	// VoteChoice defines a type for proposal vote choices.
+	VoteChoice string
+
+	// Proposal defines a DAO proposal.
+	Proposal struct {
+		id             uint64
+		status         ProposalStatus
+		definition     ProposalDefinition
+		creator        std.Address
+		record         *VotingRecord
+		statusReason   string
+		votingDeadline time.Time
+		createdAt      time.Time
+	}
+
+	// ProposalDefinition defines an interface for custom proposal definitions.
+	// These definitions define proposal content and behavior, they esentially
+	// allow the definition for different proposal types.
+	ProposalDefinition interface {
+		// Title returns the proposal title.
+		Title() string
+
+		// Body returns the proposal body.
+		// It usually contains the proposal description and other elements like proposal parameters.
+		Body() string
+
+		// VotingPeriod returns the period where votes are allowed after proposal creation.
+		// No more votes should be allowed once this period is met. It is used to calculate
+		// the voting deadline from the proposal's creationd date.
+		VotingPeriod() time.Duration
+
+		// Quorum returns the percentage of members that must vote to be able to pass a proposal.
+		// This is an optional value. DAOs use a default value when quorum is zero or invalid.
+		// Its value must be between 0 and 1, being 1 = 100% of member votes.
+		Quorum() float64
```

Comment:
What if votes have different voting power depending on the member?

Reply:
The `commondao` package wasn't implemented with voting power mechanics in mind, it started as a simple moderation DAO aimed for boards2 dApp which was eventually repurposed into common DAO. At the time the idea wasn't to use tiers or voting power, and after the repurpose we considered AtomOne's [CommonDAO spec](https://github.com/atomone-hub/genesis/blob/main/CONSTITUTION.md#section-4a-common-dao-spec) as reference of what we want to support in the short term. This specifies that each member has the same voting power.

In any case, that doesn't mean we can't change `commondao` package to support more cases if we want to aim to have a more extensive package 🚀

Reply:
By the way, the `Quorum()` method might be redundant, it's something I have to check

Reply:
Changeset c4ec22a improves tallying by making it more flexible allowing devs to optionally check quorum.
`Quorum()` has been removed and a `IsQuorumReached()` function was added in case devs want to check quorum.

> What if votes have different voting power depending on the member?

I guess that right now devs could accomplish if by using a custom index or other method to keep track of the voting power and use that in the `ProposalDefinition.Tally()`

Apart from that I created PR #3933 that explores a way to support grouping users which could allow that without the need of keeping a secondary users grouping or index.